File: netlify/functions/stats.js
const https = require('https');
const sqlite3 = require('sqlite3');
const fs = require('fs');
const path = require('path');
const os = require('os');

// Dropbox直接ダウンロードURL
const DROPBOX_URL = 'https://www.dropbox.com/scl/fi/dljhp6xzshdgvq7vqk3sz/sunsun_final_dialogue_database_proper.db?rlkey=qlf38ydm1b0n0ocsdbpjx0ih8&st=2h1nmfhq&dl=1';

let dbPath = null;

function downloadDatabase() {
  return new Promise((resolve, reject) => {
    if (dbPath && fs.existsSync(dbPath)) {
      resolve(dbPath);
      return;
    }

    dbPath = path.join(os.tmpdir(), 'sunsun_database.db');
    
    const file = fs.createWriteStream(dbPath);
    
    const request = https.get(DROPBOX_URL, (response) => {
      // リダイレクトをフォロー
      if (response.statusCode >= 300 && response.statusCode < 400 && response.headers.location) {
        const redirectUrl = response.headers.location;
        console.log('Redirected to:', redirectUrl);
        
        const redirectRequest = https.get(redirectUrl, (redirectResponse) => {
          redirectResponse.pipe(file);
          file.on('finish', () => {
            file.close();
            console.log('Database downloaded successfully, size:', fs.statSync(dbPath).size);
            resolve(dbPath);
          });
        });
        
        redirectRequest.on('error', (err) => {
          console.error('Redirect request error:', err);
          reject(err);
        });
      } else {
        response.pipe(file);
        file.on('finish', () => {
          file.close();
          console.log('Database downloaded successfully, size:', fs.statSync(dbPath).size);
          resolve(dbPath);
        });
      }
    });
    
    request.on('error', (err) => {
      console.error('Download error:', err);
      reject(err);
    });
    
    file.on('error', (err) => {
      console.error('File write error:', err);
      reject(err);
    });
  });
}

exports.handler = async (event, context) => {
  const headers = {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
    'Access-Control-Allow-Headers': 'Content-Type',
    'Content-Type': 'application/json'
  };

  // Handle OPTIONS request for CORS
  if (event.httpMethod === 'OPTIONS') {
    return {
      statusCode: 200,
      headers,
      body: ''
    };
  }

  // Handle GET request
  if (event.httpMethod === 'GET') {
    try {
      // データベースをダウンロード
      const dbFile = await downloadDatabase();
      
      return new Promise((resolve, reject) => {
        const db = new sqlite3.Database(dbFile);
        
        // 基本統計を取得
        db.get('SELECT COUNT(DISTINCT script_name) as total_scripts FROM dialogues', (err, scriptsResult) => {
          if (err) {
            reject(err);
            return;
          }
          
          db.get('SELECT COUNT(*) as total_dialogues FROM dialogues', (err, dialoguesResult) => {
            db.close();
            
            if (err) {
              reject(err);
              return;
            }
            
            const response = {
              success: true,
              data: {
                total_scripts: scriptsResult.total_scripts,
                total_dialogues: dialoguesResult.total_dialogues,
                status: 'Database loaded successfully'
              }
            };
            
            resolve({
              statusCode: 200,
              headers,
              body: JSON.stringify(response)
            });
          });
        });
      });
      
    } catch (error) {
      console.error('Error in stats handler:', error);
      
      const response = {
        success: false,
        error: error.message
      };
      
      return {
        statusCode: 500,
        headers,
        body: JSON.stringify(response)
      };
    }
  }

  // Method not allowed
  return {
    statusCode: 405,
    headers,
    body: JSON.stringify({ error: 'Method not allowed' })
  };
};

File: netlify/functions/stats.py
import json
import sqlite3
import urllib.request
import tempfile
import os
import ssl

# Dropbox直接ダウンロードURL
DROPBOX_URL = 'https://www.dropbox.com/scl/fi/dljhp6xzshdgvq7vqk3sz/sunsun_final_dialogue_database_proper.db?rlkey=qlf38ydm1b0n0ocsdbpjx0ih8&st=2h1nmfhq&dl=1'

# データベース一時ファイル
db_path = None

def download_database():
    """Dropboxからデータベースファイルをダウンロード"""
    global db_path
    
    if db_path and os.path.exists(db_path):
        return db_path
    
    try:
        # 一時ファイルを作成
        temp_fd, db_path = tempfile.mkstemp(suffix='.db')
        os.close(temp_fd)
        
        # SSL証明書検証をスキップ
        ssl_context = ssl.create_default_context()
        ssl_context.check_hostname = False
        ssl_context.verify_mode = ssl.CERT_NONE
        
        opener = urllib.request.build_opener(urllib.request.HTTPSHandler(context=ssl_context))
        urllib.request.install_opener(opener)
        
        urllib.request.urlretrieve(DROPBOX_URL, db_path)
        
        return db_path
        
    except Exception as e:
        print(f"Error downloading database: {e}")
        raise

def get_db_connection():
    """データベース接続を取得"""
    db_file = download_database()
    conn = sqlite3.connect(db_file)
    conn.row_factory = sqlite3.Row
    return conn

def main(event, context):
    """Netlify Function handler"""
    
    headers = {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type',
        'Content-Type': 'application/json'
    }
    
    # Handle OPTIONS request for CORS
    if event['httpMethod'] == 'OPTIONS':
        return {
            'statusCode': 200,
            'headers': headers,
            'body': ''
        }
    
    # Handle GET request
    if event['httpMethod'] == 'GET':
        try:
            # データベース統計を取得
            conn = get_db_connection()
            cursor = conn.cursor()
            
            # 基本統計
            cursor.execute('SELECT COUNT(DISTINCT script_name) as total_scripts FROM dialogues')
            total_scripts = cursor.fetchone()['total_scripts']
            
            cursor.execute('SELECT COUNT(*) as total_dialogues FROM dialogues')
            total_dialogues = cursor.fetchone()['total_dialogues']
            
            conn.close()
            
            response = {
                'success': True,
                'data': {
                    'total_scripts': total_scripts,
                    'total_dialogues': total_dialogues,
                    'status': 'Database loaded successfully'
                }
            }
            
            return {
                'statusCode': 200,
                'headers': headers,
                'body': json.dumps(response)
            }
            
        except Exception as e:
            print(f"Error in stats handler: {e}")
            response = {
                'success': False,
                'error': str(e)
            }
            
            return {
                'statusCode': 500,
                'headers': headers,
                'body': json.dumps(response)
            }
    
    # Method not allowed
    return {
        'statusCode': 405,
        'headers': headers,
        'body': json.dumps({'error': 'Method not allowed'})
    }